### PR TITLE
Auth: Enable retries and transaction for some db calls for auth tokens 

### PR DESCRIPTION
**Pull request by bergquist**
ref https://github.com/grafana/grafana/issues/16638

### CODE REVIEW

**Comment by bergquist** [comment]
Migrating from the old way to using `WithDbSession` I realized how bad the API is. We need to figure out a better API for this. Its very easy to write code like this
```go
var affected int64
err := srv.SQLStore.WithDbSession(ctx, func(dbSession *sqlstore.DBSession) error {
	// code

	affected, err := res.RowsAffected() 
	return err
})
```
and create an new variable in that scope instead of assigning the original one.

File: pkg/services/auth/auth_token.go [hunk 1]
```diff
@@ -80,14 +96,27 @@ func (s *UserAuthTokenService) CreateToken(userId int64, clientIP, userAgent str
...
 	hashedToken := hashToken(unhashedToken)
 	if setting.Env == setting.DEV {
 		s.log.Debug("looking up token", "unhashed", unhashedToken, "hashed", hashedToken)
 	}
 
 	var model userAuthToken
-	exists, err := s.SQLStore.NewSession().Where("(auth_token = ? OR prev_auth_token = ?) AND created_at > ? AND rotated_at > ?", hashedToken, hashedToken, s.createdAfterParam(), s.rotatedAfterParam()).Get(&model)
+	var exists bool
+	var err error
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 2]
```diff
@@ -65,7 +76,12 @@ func (s *UserAuthTokenService) CreateToken(userId int64, clientIP, userAgent str
 		SeenAt:        0,
 		AuthTokenSeen: false,
 	}
-	_, err = s.SQLStore.NewSession().Insert(&userAuthToken)
+
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 3]
```diff
@@ -35,14 +36,24 @@ func (s *UserAuthTokenService) Init() error {
...
 }
 
-func (s *UserAuthTokenService) ActiveTokenCount() (int64, error) {
-	var model userAuthToken
-	count, err := s.SQLStore.NewSession().Where(`created_at > ? AND rotated_at > ?`, s.createdAfterParam(), s.rotatedAfterParam()).Count(&model)
+func (s *UserAuthTokenService) ActiveTokenCount(ctx context.Context) (int64, error) {
+
+	var count int64
+	var err error
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 4]
```diff
@@ -100,7 +129,18 @@ func (s *UserAuthTokenService) LookupToken(unhashedToken string) (*models.UserTo
 		modelCopy := model
 		modelCopy.AuthTokenSeen = false
 		expireBefore := getTime().Add(-urgentRotateTime).Unix()
-		affectedRows, err := s.SQLStore.NewSession().Where("id = ? AND prev_auth_token = ? AND rotated_at < ?", modelCopy.Id, modelCopy.PrevAuthToken, expireBefore).AllCols().Update(&modelCopy)
+
+		var affectedRows int64
+		err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 5]
```diff
@@ -116,7 +156,17 @@ func (s *UserAuthTokenService) LookupToken(unhashedToken string) (*models.UserTo
 		modelCopy := model
 		modelCopy.AuthTokenSeen = true
 		modelCopy.SeenAt = getTime().Unix()
-		affectedRows, err := s.SQLStore.NewSession().Where("id = ? AND auth_token = ?", modelCopy.Id, modelCopy.AuthToken).AllCols().Update(&modelCopy)
+
+		var affectedRows int64
+		err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 6]
```diff
@@ -183,12 +233,21 @@ func (s *UserAuthTokenService) TryRotateToken(token *models.UserToken, clientIP,
 			rotated_at = ?
 		WHERE id = ? AND (auth_token_seen = ? OR rotated_at < ?)`
 
-	res, err := s.SQLStore.NewSession().Exec(sql, userAgent, clientIP, s.SQLStore.Dialect.BooleanStr(true), hashedToken, s.SQLStore.Dialect.BooleanStr(false), now.Unix(), model.Id, s.SQLStore.Dialect.BooleanStr(true), now.Add(-30*time.Second).Unix())
+	var affected int64
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 7]
```diff
@@ -199,14 +258,20 @@ func (s *UserAuthTokenService) TryRotateToken(token *models.UserToken, clientIP,
...
 	if token == nil {
 		return models.ErrUserTokenNotFound
 	}
 
 	model := userAuthTokenFromUserToken(token)
 
-	rowsAffected, err := s.SQLStore.NewSession().Delete(model)
+	var rowsAffected int64
+	var err error
+	err = s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 8]
```diff
@@ -221,55 +286,71 @@ func (s *UserAuthTokenService) RevokeToken(token *models.UserToken) error {
...
 }
 
-func (s *UserAuthTokenService) RevokeAllUserTokens(userId int64) error {
-	sql := `DELETE from user_auth_token WHERE user_id = ?`
-	res, err := s.SQLStore.NewSession().Exec(sql, userId)
-	if err != nil {
-		return err
-	}
+func (s *UserAuthTokenService) RevokeAllUserTokens(ctx context.Context, userId int64) error {
+	return s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 9]
```diff
@@ -221,55 +286,71 @@ func (s *UserAuthTokenService) RevokeToken(token *models.UserToken) error {
...
-	}
-
-	if !exists {
-		return nil, models.ErrUserTokenNotFound
-	}
+func (s *UserAuthTokenService) GetUserToken(ctx context.Context, userId, userTokenId int64) (*models.UserToken, error) {
 
 	var result models.UserToken
-	token.toUserToken(&result)
+	err := s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

File: pkg/services/auth/auth_token.go [hunk 10]
```diff
@@ -221,55 +286,71 @@ func (s *UserAuthTokenService) RevokeToken(token *models.UserToken) error {
...
-	}
+func (s *UserAuthTokenService) GetUserTokens(ctx context.Context, userId int64) ([]*models.UserToken, error) {
 
 	result := []*models.UserToken{}
-	for _, token := range tokens {
-		var userToken models.UserToken
-		token.toUserToken(&userToken)
-		result = append(result, &userToken)
-	}
+	err := s.SQLStore.WithDbSession(context.Background(), func(dbSession *sqlstore.DBSession) error {
```
Comment (marefr) [inline pkg/services/auth/auth_token.go]: s/context.Background()/ctx

**Review by marefr — CHANGES_REQUESTED** [review]
Just to be clear this doesn't add any retries for sqlite right?

You did a very thorough work getting context in everywhere, but in auth_token.go you forget to use the provided ctx as argument to `WithDbSession` :)

**Comment by torkelo** [comment]
Just to understand the db locked issue and how this solves it. Was it caused, you suspect, by opening & not closing sessions? 

Wonder if it would be possible to open a session in the middleware at the start and store it in the Request context, and the pass context down to sql layer and have the withDbSession reuse session in the context object. And have the middleware close the session on request completion.

**Comment by bergquist** [comment]
Fixed the `context.Background` issues and switched to using transactions (inc retries) for rotation. I was considering adding retries for all WithDbSession requests as well but Im not sure about it. It could potentially create more requests to the db if it's exhausted. 

@torkelo yesterday I didn't like that idea but I might have changed my mind. We still need some way of managing session background tasks. Let's discuss that in another issue. I think we need to improve this soon as more people will be working in the backend.

File: pkg/services/quota/quota.go [hunk 11]
```diff
@@ -20,6 +22,17 @@ func (qs *QuotaService) Init() error {
...
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+	fmt.Println("kasdgfkjhgsdafkjhgasdkjhf")
+
```
Comment (marefr) [inline pkg/services/quota/quota.go]: Please remove this :smile:

**Review by marefr — CHANGES_REQUESTED** [review]
Everything looks good besides some dummy code that seems to have been sneaked in, see comment.

File: pkg/services/quota/quota.go (hunk 11 above)
Comment (bergquist) [reply pkg/services/quota/quota.go]: haha

**Review by marefr — APPROVED** [review]
:tada:

**Pull request merged**